client_unix.go

```diff
@@ -5,15 +5,66 @@ package containerd
 import (
 	"fmt"
 	"net"
+	"os"
 	"strings"
+	"syscall"
 	"time"
+
+	"github.com/pkg/errors"
 )
 
-func dialer(address string, timeout time.Duration) (net.Conn, error) {
+func isNoent(err error) bool {
+	if err != nil {
+		if nerr, ok := err.(*net.OpError); ok {
+			if serr, ok := nerr.Err.(*os.SyscallError); ok {
+				if serr.Err == syscall.ENOENT {
+					return true
+				}
+			}
+		}
+	}
+	return false
+}
+
+func Dialer(address string, timeout time.Duration) (net.Conn, error) {
+	var (
+		c     net.Conn
+		err   error
+		stopC = make(chan struct{})
+		synC  = make(chan struct{})
```

Comment:
You could have this chan return the `net.Conn` and remove some of the sync stuff at the end.  Just have a select accepting the conn and one on the timeout